network/src/internal/peer_book.rs

```diff
@@ -0,0 +1,267 @@
+// Copyright (C) 2019-2020 Aleo Systems Inc.
+// This file is part of the snarkOS library.
+
+// The snarkOS library is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// The snarkOS library is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with the snarkOS library. If not, see <https://www.gnu.org/licenses/>.
+
+use crate::internal::PeerInfo;
+use snarkos_errors::network::ServerError;
+use snarkos_metrics::Metrics;
+use snarkos_models::{algorithms::LoadableMerkleParameters, objects::Transaction};
+use snarkos_storage::Ledger;
+
+use serde::{Deserialize, Serialize};
+use std::{collections::HashMap, net::SocketAddr};
+
+/// An data structure for tracking and indexing the history of
+/// all connected and disconnected peers to the node.
+#[derive(Clone, Debug, Serialize, Deserialize)]
```

Comment:
as before, deriving `Clone` on objects containing collections is usually undesirable due to deep-cloning; if `PeerBook` is expected to be wrapped in an `Arc`, it doesn't need to be `Clone`able - the `Arc` can still be cloned